packages/arb-avm-cpp/avm_values/include/avm_values/buffer.hpp

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright 2019-2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef buffer_hpp
+#define buffer_hpp
+
+#include <memory>
+#include <vector>
+#include <cstdint>
+#include <iostream>
+#include <avm_values/bigint.hpp>
+
+inline uint64_t calc_len(int h) {
+    if (h == 0) {
+        return 1024;
+    }
+    return 128*calc_len(h-1);
+}
+
+struct Packed {
+    uint256_t hash;
+    int size; // total size
+    int packed; // packed levels
+};
+
+Packed zero_packed(int sz);
+
+class RawBuffer {
+   private:
+    bool is_leaf;
+    int level;
```

Comment:
Is `is_leaf` true if and only if `level == 0`. If so `is_leaf` should be eliminated to minimize how much consistency we need to enforce.